contracts/oracles/Oracle.sol

```diff
@@ -111,25 +119,30 @@ contract Oracle is Ownable, ReentrancyGuard {
 
     /// @notice Vote an Event result which requires BOT payment.
     /// @param _eventResultIndex The Event result which is being voted on.
-    function voteResult(uint8 _eventResultIndex) 
+    /// @param _botAmount The amount of BOT to use for voting.
+    function voteResult(uint8 _eventResultIndex, uint256 _botAmount) 
         public 
-        payable 
         validResultIndex(_eventResultIndex) 
     {
-        require(msg.value > 0);
+        require(_botAmount > 0);
+        require(totalStakeContributed.add(_botAmount) <= maxStakeContribution);
         require(block.number >= eventBettingEndBlock);
         require(block.number < decisionEndBlock);
         require(!participants[msg.sender].didSetResult);
+        ERC20 token = ERC20(addressManager.bodhiTokenAddress());
+        require(token.allowance(msg.sender, address(this)) >= _botAmount);
 
         Participant storage participant = participants[msg.sender];
-        participant.stakeContributed = participant.stakeContributed.add(msg.value);
+        participant.stakeContributed = participant.stakeContributed.add(_botAmount);
```

Comment:
We are using the `SafeMath` library for math functions to properly handle overflow and such. 

I see what you are saying with the race condition. To safeguard, I've moved `participant.didSetResult = true;` to the top of first state change. This way if there is a race condition, it will fail because of the `require(!participants[msg.sender].didSetResult)` check. 

Although, this should never really hit a race condition due to a reentrancy attack as its impossible in this case. There is no way to hit the fallback function to recall our contract.

Reply:
Oh also for your suggestion for using set(), the logic is basically setting the value:
```
eventResults[_eventResultIndex].votedBalance = eventResults[_eventResultIndex].votedBalance.add(_botAmount);
```
is saying take the old votedBalance, add the _botAmount, then set it to eventResults[_eventResultIndex].votedBalance.

Either way, the didSetResult change should prevent any race conditions.